seps/sep-0002.md

```diff
@@ -0,0 +1,102 @@
+# Use-variants: for packages with conflicting libraries
+
+## The problem
+
+The `intel-oneapi-mkl` package is installed with many libraries. Some
+of these libraries define the same symbols but differ in their
+implementation. For example there is a set of libraries built with
+threading support, and another set which offers the same
+functionality but are built without threading support.
+
+Right now, the only mechanism Spack offers to choose between these
+libraries is a variant: you could add a `threading` variant and then
+choose which libraries to supply to dependents based on the value
+of that variant. The problem with a variant is that different
+variant values generate distinct installations of a package: we would
+install `intel-oneapi-mkl` once for each choice of `threading`, even
+though the installations would be the same.
+
+See also: https://github.com/spack/spack/discussions/22749
+
+## Proposed Changes
+
+* Modify variant declarations with a flag to indicate that the variant
+  determines how the package is used rather than how it is installed
+* During concretization, this variant can be treated like any other
+  variant.
+* After concretization, the variant is stripped from the node and
+  stored on the edge connecting the node to its parent.
```

Comment:
I think this is the optimal way of proceeding with the original concretizer. It should work also with the ASP-based solver, but I think in that case it would be better to model the semantics in the logic program (i.e. in `concretize.lp`). @tgamblin Do you have already some thoughts on it?

Reply:
I think this doesn't handle all the cases yet -- in particular, we discussed wanting to be able to constrain how some use variants were used but not others.  If we keep the use-variant on the spec, and only strip it off after, then we're forcing all of them to be unified.  But I think you could have cases where one dependent uses, e.g., a boost library and another does not -- and that could be perfectly fine.  This spec doesn't yet seem to cover how to constrain use variants .. maybe @becker33 could take a stab at that as it was one of his examples.

Reply:
I also discuss Boost as a use case and came to the conclusion that it is resolvable in a simpler manner (which I describe there). I intended to propose that use variants (for other readers, that is variants with `interface=True`) should be treated exactly like variants during concretization.